### PR TITLE
lazyload - breakpoints

### PR DESCRIPTION
**Pull request by DavidCramer**

### CODE REVIEW

File: php/class-responsive-breakpoints.php [hunk 1]
```diff
@@ -0,0 +1,336 @@
...
+			$palcehold_str             = Api::generate_transformation_string( $placehold_transformations );
+			$placeholder               = str_replace( $original_string, $palcehold_str, $src );
+
+			$tag_element['atts']['src'] = $placeholder;
+			if ( ! empty( $settings['lazy_preloader'] ) ) {
+				$tag_element['atts']['data-placeholder'] = $placeholder;
+			}
+		}
+
+		if ( ! empty( $settings['lazy_preloader'] ) ) {
```
Comment (pereirinha) [inline php/class-responsive-breakpoints.php:187]: @DavidCramer 

This is a reminder for introducing the controls discussed for the preloader:
+ Predominant color;
+ Color picker;
+ Fade animation;

Also the preloader should always be there.

File: php/class-responsive-breakpoints.php [hunk 2]
```diff
@@ -0,0 +1,336 @@
...
+
+		unset( $tag_element['atts']['loading'] );
+		$tag_element['atts']['decoding']   = 'async';
+		$tag_element['atts']['data-width'] = $meta['width'];
+
+		return $tag_element;
+	}
+
+	/**
+	 * Check if component is active.§
```
Comment (pereirinha) [inline php/class-responsive-breakpoints.php:201]: @DavidCramer 

Remove the alien character :)

File: php/class-responsive-breakpoints.php [hunk 3]
```diff
@@ -0,0 +1,336 @@
...
+	}
+
+	/**
+	 * Enqueue Assets
+	 */
+	public function enqueue_assets() {
+
+		wp_enqueue_script( 'cld-responsive-breakpoints' );
+		$breakpoints = $this->media->get_settings()->get_value( 'image_display' );
+		wp_add_inline_script( 'cld-responsive-breakpoints', 'var CLDLB = ' . wp_json_encode( $breakpoints ), 'before' );
```
Comment (pereirinha) [inline php/class-responsive-breakpoints.php:228]: @DavidCramer 

What does `CLDLB` stands for?

File: php/class-responsive-breakpoints.php [hunk 4]
```diff
@@ -0,0 +1,336 @@
...
+			array(
+				'type'      => 'group',
+				'condition' => array(
+					'use_lazy_loading' => true,
+				),
+				array(
+					'type'        => 'number',
+					'title'       => __( 'Lazy loading threshold', 'cloudinary' ),
+					'slug'        => 'lazy_threshold',
+					'description' => __( ' The threshold', 'cloudinary' ),
```
Comment (pereirinha) [inline php/class-responsive-breakpoints.php:272]: @DavidCramer 

There's an space ` ` in the beginning of the string, and it's repeated a few more times.

File: ui-definitions/settings-image.php [hunk 5]
```diff
@@ -87,6 +87,9 @@
 				),
 				array(
 					'type' => 'group',
+					'title'        => __( 'Image Display', 'cloudinary' ),
```
Comment (pereirinha) [inline ui-definitions/settings-image.php:90]: @DavidCramer

let's fix the alignments here.

File: php/class-responsive-breakpoints.php [hunk 6]
```diff
@@ -0,0 +1,336 @@
...
+use Cloudinary\Component\Assets;
+use Cloudinary\Component\Setup;
+use Cloudinary\Connect\Api;
+
+/**
+ * Class Responsive_Breakpoints
+ *
+ * @package Cloudinary
+ */
+class Responsive_Breakpoints implements Setup, Assets {
```
Comment (pereirinha) [inline php/class-responsive-breakpoints.php:19]: @DavidCramer 

It seems that this class is addressing both Responsive Breakpoints and Lazy Loading concepts.
What do you think of separate these concerns?

File: php/class-media.php [hunk 7]
```diff
@@ -1935,7 +1935,7 @@ public function get_breakpoint_options( $attachment_id ) {
 		$settings    = $this->settings->get_setting( self::MEDIA_SETTINGS_SLUG )->get_value();
 
 		if ( 'on' === $settings['enable_breakpoints'] && wp_attachment_is_image( $attachment_id ) ) {
-			$meta = wp_get_attachment_metadata( $attachment_id );
+			$meta = wp_get_attachment_metadata( $attachment_id, true );
```
Comment (pereirinha) [inline php/class-media.php:1938]: @DavidCramer 

We do still have other instances where we use the filtered `wp_get_attachment_metadata`. Is there as opportunity to normalize this, or those cases are actually supposed to factor in any filters?

**Review by pereirinha — COMMENTED** [review]
Hey @DavidCramer 

I left you some comments.
Can you please weigh in.

Thanks.

File: js/src/responsive-breakpoints.js [hunk 8]
```diff
@@ -0,0 +1,144 @@
+const ResponsiveBreakpoints = {
+	density: window.devicePixelRatio ? window.devicePixelRatio : 'auto',
```
[the remark this answers is not in the document]
Comment (DavidCramer) [reply js/src/responsive-breakpoints.js:2]: @pereirinha - this is the default detected. see method `_getDensity()`

File: js/src/responsive-breakpoints.js [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+		const width = this.scaleSize(
+			image.originalWidth,
+			image.width,
+			this.config.bytes_step
+		);
+		const newSize = 'w_' + width + ',dpr_' + this.density;
+		return image.dataset.src
+			.replace( '--size--', newSize )
+			.replace( '/--placehold--', '' )
+			.replace( 'q_auto', this.getQuality() );
```
[the remark this answers is not in the document]
Comment (DavidCramer) [reply js/src/responsive-breakpoints.js]: @pereirinha - corrected.

File: php/class-connect.php [hunk 10]
```diff
@@ -857,7 +857,7 @@ public function settings() {
 						'link_text' => __( 'See Examples', 'cloudinary' ),
 					),
 				),
-				'connect' => array(
+				'connection' => array(
```
[the remark this answers is not in the document]
Comment (DavidCramer) [reply php/class-connect.php]: I've corrected it. The one that cause the duplication, is now corrected, which doesn't cause any issues.

**Review by pereirinha — APPROVED** [review]
Thanks for the great stuff @DavidCramer.

**Pull request merged**